### PR TITLE
DNS-SD: Set hostname to be mac.

### LINKED ISSUE #5270: Fix MDNS hostname
#### Problem
MDNS hostname is incorrect - per spec it should be the 48-bit mac or EUI-64. Right now it matches the instance name.

 #### Proposed Solution
Set the hostname to be the mac address for wifi. Adding another issue to track thread.

### LINKED ISSUE #5271: Fix MDNS hostname for thread
#### Problem
Per spec, mdns hostname should be EUI-64 for thread. Right now it matches the instance name.

 #### Proposed Solution
Fix the hostname for thread devices.

### PR DESCRIPTION
**Pull request by cecille**
#### Problem
Current hostname is set to be the same as the instance name, but should actually be the MAC address.
Please see spec section 4.2.1 commissioning discovery and 4.2.2 operational discovery

 #### Summary of Changes
Adds mac and hostname to rendezvous parameters and sets these in the mdns advertiser impl.

Test: avahi-discover on M5 stack using operational and commissioning discovery - hostname is now correct.

 Fixes #5271, fixes #5270

### CODE REVIEW

File: src/lib/mdns/Advertiser.h [hunk 1]
```diff
@@ -51,11 +52,37 @@ class BaseAdvertisingParams
         return *reinterpret_cast<Derived *>(this);
     }
     bool IsIPv4Enabled() const { return mEnableIPv4; }
+    Derived & SetMac(uint8_t * mac, size_t size)
```
Comment (andy31415) [inline src/lib/mdns/Advertiser.h:55]: nit: consider using chip::Span to represent a buffer+length pair

File: src/lib/mdns/Advertiser.h [hunk 2]
```diff
@@ -51,11 +52,37 @@ class BaseAdvertisingParams
...
+    size_t GetMacSize() const { return macSize; }
+    const char * GetHostname() const { return hostname; }
 
 private:
-    uint16_t mPort   = CHIP_PORT;
-    bool mEnableIPv4 = true;
-};
+    static constexpr size_t kMaxMacSize = 8;
+    // 2-byte ascii for 8-byte mac + 1 for null
+    static constexpr size_t kMaxHostnameSize = kMaxMacSize * 2;
```
Comment (andy31415) [inline src/lib/mdns/Advertiser.h:79]: should this contain the "+1 for null" part or move the comment and calculation to the hosstname array?

File: src/lib/mdns/Advertiser.h [hunk 3]
```diff
@@ -51,11 +52,37 @@ class BaseAdvertisingParams
...
+    uint16_t mPort                           = CHIP_PORT;
+    bool mEnableIPv4                         = true;
+    uint8_t mMac[kMaxMacSize]                = {};
+    size_t macSize                           = kMaxMacSize;
+    char hostname[kMaxHostnameSize + 1]      = "000000000000";
```
Comment (andy31415) [inline src/lib/mdns/Advertiser.h:84]: standard naming seems to be m-prefixed: `mMacSize` and `mHostName`

File: src/lib/mdns/Advertiser.h (hunk 1 above)
Comment (cecille) [reply src/lib/mdns/Advertiser.h]: Added in next commit - thanks. Still finding all the chip goodies in the codebase :)

File: src/lib/mdns/Advertiser.h (hunk 2 above)
Comment (cecille) [reply src/lib/mdns/Advertiser.h]: Done.

File: src/lib/mdns/Advertiser.h (hunk 3 above)
Comment (cecille) [reply src/lib/mdns/Advertiser.h]: done. Thanks.

File: src/app/server/Mdns.cpp [hunk 4]
```diff
@@ -64,6 +65,26 @@ NodeId GetCurrentNodeId()
...
+    {
+        ChipLogDetail(Discovery, "Using Thread MAC for hostname.");
+        return chip::ByteSpan(mac, 8);
+    }
+    if (DeviceLayer::ConfigurationMgr().GetPrimaryWiFiMACAddress(mac) == CHIP_NO_ERROR)
+    {
+        ChipLogDetail(Discovery, "Using wifi MAC for hostname");
+        return chip::ByteSpan(mac, 6);
+    }
+    ChipLogError(Discovery, "Wifi mac not known. Using a default.");
```
Comment (asakhart) [inline src/app/server/Mdns.cpp:82]: What about Ethernet devices, we have some that folks are testing with?

File: src/app/server/Mdns.cpp [hunk 5]
```diff
@@ -64,6 +65,26 @@ NodeId GetCurrentNodeId()
     return chip::kTestDeviceNodeId;
 }
 
+// Requires an 8-byte mac to accommodate thread.
+chip::ByteSpan FillMAC(uint8_t mac[8])
+{
+    memset(mac, 0, 8);
+    if (DeviceLayer::ConfigurationMgr().GetPrimary802154MACAddress(mac) == CHIP_NO_ERROR)
```
Comment (LuDuda) [inline src/app/server/Mdns.cpp:72]: I'm a little bit confused here. A few days ago #5343 was merged which added support for setting hostname. Please look [here](CHIP_ERROR DiscoveryImplPlatform::SetupHostname()). Is it a different implementation?

Side note: The specification defines the usage of EUI-64 for Thread, so we should use `_GetFactoryAssignedEUI64` instead.

Comment (cecille) [reply src/app/server/Mdns.cpp:72]: Crossed streams a bit - I didn't see that go in. But without this change, the hostname I'm getting back for non-thread devices is definitely incorrect even with #5343. Let me see if I can merge these two implementations a bit more nicely

Comment (kkasperczyk-no) [reply src/app/server/Mdns.cpp:72]: That's true, WiFi implementation from #5343 may not be correct, as I didn't modify existing solution. I just added missing part of creating hostname for Thread devices.

File: src/lib/mdns/Advertiser_ImplMinimalMdns.cpp [hunk 6]
```diff
@@ -411,7 +411,7 @@ CHIP_ERROR AdvertiserMinMdns::Advertise(const OperationalAdvertisingParameters &
 
     FullQName operationalServiceName = AllocateQName("_chip", "_tcp", "local");
     FullQName operationalServerName  = AllocateQName(uniqueName, "_chip", "_tcp", "local");
-    FullQName serverName             = AllocateQName(uniqueName, "local");
+    FullQName serverName             = AllocateQName(params.GetHostname(), "local");
```
Comment (kkasperczyk-no) [inline src/lib/mdns/Advertiser_ImplMinimalMdns.cpp:414]: If you decided to create more generic method for building host name (what's really great), please use it also in the [Discovery_ImplPlatform.cpp](https://github.com/project-chip/connectedhomeip/blob/master/src/lib/mdns/Discovery_ImplPlatform.cpp#L149) to cover `mdns = platform` case and not only the `mdns = minimal` one.

Comment (kkasperczyk-no) [reply src/lib/mdns/Advertiser_ImplMinimalMdns.cpp:414]: That's crucial for Thread devices, as current SRP implementation provides kind of backend for mDNS only in the platform case, so Thread devices are not using minimal mDNS.

File: src/lib/mdns/Advertiser.h [hunk 7]
```diff
@@ -51,11 +57,34 @@ class BaseAdvertisingParams
...
 private:
-    uint16_t mPort   = CHIP_PORT;
-    bool mEnableIPv4 = true;
-};
+    uint16_t mPort                   = CHIP_PORT;
+    bool mEnableIPv4                 = true;
+    uint8_t mMacStorage[kMaxMacSize] = {};
+    chip::ByteSpan mMac              = chip::ByteSpan(mMacStorage, kMaxMacSize);
+    // Add 1 for the end nullptr.
+    char mHostname[kMaxHostnameSize + 1] = "000000000000";
```
Comment (kkasperczyk-no) [inline src/lib/mdns/Advertiser.h:86]: Why is it "000000000000"? The array is 17 bytes size and there are 12 characters in the string. Wouldn't be enough just to set `/0`, as a first character to mark array empty?

Comment (cecille) [reply src/lib/mdns/Advertiser.h:86]: wifi mac is 6 bytes, which translates to 12 ascii chars. I figured all zeros would be a good way to indicate that something was not correctly initialized. I'm not sure what a blank hostname will do to the mdns clients - you'll end up with a name that's just .local, which could potentially cause issues. 

I can add a comment to that effect. If you have a strong preference for empty, I'm fine to go that route too.

File: src/app/server/Mdns.cpp (hunk 4 above)
Comment (cecille) [reply src/app/server/Mdns.cpp:82]: That is a good question - Spec specifies mac for ethernet too, but it's probably in a different config. I'll take a look and add that in. Thanks.

That does bring up another interesting question though - for dual devices, what should we prefer? With the code as it is right now, border routers will prefer the EUI-64, which may not actually be what we want. Ditto for dual ethernet / wifi devices - if the ethernet mac comes from a different config, we may want a way to test what's actually connected. Currently, it's just an ID so it doesn't actually matter too much for functionality, and the spec doesn't have a stated preference.

File: src/lib/mdns/Advertiser_ImplMinimalMdns.cpp (hunk 6 above)
Comment (cecille) [reply src/lib/mdns/Advertiser_ImplMinimalMdns.cpp:414]: Can do - I'll take a look today. Thanks.

File: src/lib/mdns/Advertiser.h (hunk 7 above)
Comment (kkasperczyk-no) [reply src/lib/mdns/Advertiser.h:86]: Honestly I don't know how the empty host name will be treated by the mDNS, but I guess that is doesn't have much sense, so maybe when hostname could not be assigned we should assert? For example SRP doesn't really care about what host name contains, so assigning all zeros will result in creating record in SRP server for host: 000000000000.local. Imagine that for some reason assigning host name will fail for multiple devices and all of them will try to register as the same host called 000000000000.

Comment (cecille) [reply src/lib/mdns/Advertiser.h]: Code has changed - this is now removed.

File: src/lib/mdns/Advertiser_ImplMinimalMdns.cpp (hunk 6 above)
Comment (cecille) [reply src/lib/mdns/Advertiser_ImplMinimalMdns.cpp]: These should now be the same - PTAL.

**Review by andy31415 — APPROVED** [review]

**Review by mspang — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Comment by andy31415** [comment]
@kkasperczyk-no @Damian-Nordic @woody-apple @bzbarsky-apple ?

**Review by Damian-Nordic — APPROVED** [review]

**Pull request merged**